proxy/src/connection.rs

```diff
@@ -185,23 +191,64 @@ impl Future for Connecting {
     fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
         loop {
             *self = match self {
-                Connecting::Plaintext { connect, tls } => {
+                Connecting::Plaintext { addr, connect, tls } => {
                     let plaintext_stream = try_ready!(connect.poll());
                     set_nodelay_or_warn(&plaintext_stream);
                     match tls.take().expect("Polled after ready") {
                         Conditional::Some(config) => {
-                            let upgrade_to_tls = tls::Connection::connect(
+                            let upgrading = tls::Connection::connect(
                                 plaintext_stream, &config.identity, config.config);
-                            Connecting::UpgradeToTls(upgrade_to_tls)
+                            Connecting::UpgradeToTls { addr: *addr, upgrading }
                         },
                         Conditional::None(why) => {
                             return Ok(Async::Ready(Connection::plain(plaintext_stream, why)));
                         },
                     }
                 },
-                Connecting::UpgradeToTls(upgrading) => {
-                    let tls_stream = try_ready!(upgrading.poll());
-                    return Ok(Async::Ready(Connection::tls(tls_stream)));
+                Connecting::UpgradeToTls { addr, upgrading } => {
+                    match upgrading.poll() {
+                        Ok(Async::NotReady) => return Ok(Async::NotReady),
+                        Ok(Async::Ready(tls_stream)) => {
+                            let conn = Connection::tls(tls_stream);
+                            return Ok(Async::Ready(conn));
+                        },
+                        Err(e) => {
+                            // Did the TLS handshake fail, or did something
+                            // else go wrong?
+                            let handshake_err = if let Some(tls_error) = e
+                                .get_ref()
+                                .and_then(|e| e.downcast_ref::<tls::Error>())
+                            {
+                                warn!(
+                                    "TLS handshake with {:?} failed: {}\
+                                     -> falling back to plaintext",
+                                    addr, tls_error,
+                                );
```

Comment:
To be conservative in favor of keeping things working (vs. being secure), I think we should just do the fallback on any error, not just TLS errors.